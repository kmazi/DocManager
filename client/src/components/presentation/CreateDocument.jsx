import React from 'react';
import PropTypes from 'prop-types';
import $ from 'jquery';
import Alert from 'sweetalert2';
import TinyMCE from 'react-tinymce';

let documentBody = '';

/**
 * Creates a document object from the form
 * @param {number} userId - the userId that creating the document
 * @return {object} returns the form value to use in creting the document
 */
const setFormValue = (userId) => {
  const title = $('#docform input[name=title]').val();
  const body = documentBody;
  const access = $('#docform input[name=group1]:checked').val();
  const token = localStorage.getItem('docmanagertoken');
  const formValue = { title, body, access, userId, token };
  return formValue;
};

/**
 * Submits a completed form to create a document
 * @param {object} event - contains information
 * about the element raising the event
 * @param {func} documentCreation - a function that handles
 * the event to create documents
 * @param {object} formValue - the value of the form used to create document
 * @return {null} returns null
 */
const createDocument = (event, documentCreation, formValue) => {
  event.preventDefault();
  if (formValue.token && formValue.title !== '' && formValue.body !== '') {
    documentCreation(formValue).then((res) => {
      if (res.status === 'successful') {
        Alert({
          title: res.status,
          text: 'Your document has been created',
          type: 'success',
          confirmButtonText: 'ok'
        });
        document.getElementById('form').reset();
        documentBody = '';
      } else {
        Alert({
          title: res.status,
          text: res.message,
          type: 'error',
          confirmButtonText: 'ok'
        });
      }
    });
  } else {
    Alert({
      title: 'Incomplete form',
      text: 'Fill the form completely before submitting to the server',
      type: 'error',
      confirmButtonText: 'ok'
    });
  }
};

const handleEditorChange = (event) => {
  documentBody = event.target.getContent();
};
/**
 * Component that renders the form to create a document
 * @param {object} an object containing properties for creating a document
 * @return {object} returns the html to render
 */
const CreateDocument = ({ documentCreation, userId, roleType }) => (
  <form id="form" action="">
    <div id="docform">
      <input type="text" placeholder="title" name="title" />
      <TinyMCE
        content=""
        config={{
          plugins: 'autolink link image lists print preview',
          toolbar: 'undo redo | bold italic | alignleft aligncenter alignright',
          height: 280,
          width: '100%',
          browser_spellcheck: true,
        }}
        onChange={handleEditorChange}
      /><br />
      <div>
        Access level:
      <input
        className="with-gap"
        name="group1"
        type="radio"
        id="public"
        value="Public"
      />
        <label id="docrole" htmlFor="public">Public</label>
        <input
          className="with-gap"
          name="group1"
          value={roleType}
          type="radio"
          id="role"
        />
        <label htmlFor="role">{roleType}</label>
        <input
          className="with-gap"
          name="group1"
          type="radio"
          value="Private"
          id="private"
        />
        <label htmlFor="private">Private</label>
        <button
          className="btn waves-effect waves-light right"
          type="submit"
          name="action"
          onClick={(event) => {
            const formValue = setFormValue(userId);
            createDocument(event, documentCreation, formValue);
          }}
        >create
    </button>
      </div>
    </div>
  </form>
);

CreateDocument.propTypes = {
  documentCreation: PropTypes.func.isRequired,
  userId: PropTypes.number.isRequired,
  roleType: PropTypes.string.isRequired,
};

export default CreateDocument;
