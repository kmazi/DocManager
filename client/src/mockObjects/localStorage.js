/**
 * Mocks localstorage
 */
class LocalStorageMock {
  /**
   * the constructor
   */
  constructor() {
    this.store = {};
  }
/**
 * clears the localstorage
 * @return {void} returns null
 */
  clear() {
    this.store = {};
  }

/**
 * get item in the localstorage
 * @param {string} key - The key to the item in the localstorage
 * @return {string} returns the key to the item in localstorage
 */
  getItem(key) {
    return this.store[key];
  }
/**
 * set the value of an item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @param {string} value - The value of the item in the localstorage
 * @return {null} returns null
 */
  setItem(key, value) {
    this.store[key] = value;
  }

  /**
 * remove item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @return {void} returns nothing
 */
  removeItem(key) {
    delete this.store[key];
  }
}

global.localStorage = new LocalStorageMock();
export default global.localStorage;
