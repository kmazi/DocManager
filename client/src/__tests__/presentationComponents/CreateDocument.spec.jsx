import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import CreateDocument from '../../components/presentation/CreateDocument';
import '../../mockObjects/localStorage';

describe('The CreateDocument component:', () => {
  const response = { status: 'successful' };
  const props = {
    documentCreation: jest.fn(() => Promise.resolve(response)),
    roleType: 'Learning',
    userId: 1,
  };
  test('should render correctly', () => {
    const enzymeWrapper = shallow(<CreateDocument {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test(`should not fire create Document when
  the form is empty`,
    () => {
      const event = {
        preventDefault: () => 'Hello',
      };
      const setFormValue = jest.fn();
      setFormValue.mockReturnValueOnce({
        title: 'Hello',
        body: 'Hi',
      });
      const enzymeWrapper = shallow(<CreateDocument {...props} />);
      enzymeWrapper.find('#docform button[name="action"]')
      .props().onClick(event);
      expect(props.documentCreation.mock.calls.length).toBe(0);
    });

  test('should update the text entered when text in TinyMCE changes',
    () => {
      const event = {
        target: { getContent: () => 'hello', },
      };
      const handleEditorChange = jest.fn();
      const enzymeWrapper = shallow(<CreateDocument {...props} />);
      enzymeWrapper.find('TinyMCE').props().onChange(event);
      expect(handleEditorChange.mock.calls.length).not.toBeGreaterThan(1);
    });
});
