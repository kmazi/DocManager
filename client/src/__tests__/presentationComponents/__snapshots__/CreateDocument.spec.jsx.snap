// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CreateDocument component: should render correctly 1`] = `
<form
  action=""
  id="form"
>
  <div
    id="docform"
  >
    <input
      name="title"
      placeholder="title"
      type="text"
    />
    <TinyMCE
      config={
        Object {
          "browser_spellcheck": true,
          "height": 280,
          "plugins": "autolink link image lists print preview",
          "toolbar": "undo redo | bold italic | alignleft aligncenter alignright",
          "width": "100%",
        }
      }
      content=""
      onChange={[Function]}
    />
    <br />
    <div>
      Access level:
      <input
        className="with-gap"
        id="public"
        name="group1"
        type="radio"
        value="Public"
      />
      <label
        htmlFor="public"
        id="docrole"
      >
        Public
      </label>
      <input
        className="with-gap"
        id="role"
        name="group1"
        type="radio"
        value="Learning"
      />
      <label
        htmlFor="role"
      >
        Learning
      </label>
      <input
        className="with-gap"
        id="private"
        name="group1"
        type="radio"
        value="Private"
      />
      <label
        htmlFor="private"
      >
        Private
      </label>
      <button
        className="btn waves-effect waves-light right"
        name="action"
        onClick={[Function]}
        type="submit"
      >
        create
      </button>
    </div>
  </div>
</form>
`;
